ql/src/java/org/apache/hadoop/hive/ql/optimizer/GroupingSetOptimizer.java

```diff
@@ -0,0 +1,362 @@
+
+      List<Integer> groupingSetKeys = listGroupingSetKeyPositions(gby.getConf().getListGroupingSets());
+      Set<String> candidates = new HashSet<>();
+      for (Integer groupingSetKeyPosition: groupingSetKeys) {
+        ExprNodeDesc key = gby.getConf().getKeys().get(groupingSetKeyPosition);
+
+        if (key instanceof ExprNodeColumnDesc) {
+          candidates.add(((ExprNodeColumnDesc) key).getColumn());
+        }
+      }
+      candidates.retainAll(colNamesInSignature);
+
+      List<ColStatistics> columnStatistics = new ArrayList<>(parentOp.getStatistics().getColumnStats());;
+      columnStatistics.sort(new Comparator<ColStatistics>() {
+        @Override
+        public int compare(ColStatistics o1, ColStatistics o2) {
+          if (o1.getCountDistint() == o2.getCountDistint()) {
+            return 0;
+          } else {
+            // sort in reversed order, i.e., the largest comes the first
+            return (o1.getCountDistint() < o2.getCountDistint()) ? 1 : -1;
+          }
+        }
+      });
+
+      String partitionCol = null;
+      for (ColStatistics col: columnStatistics) {
+        String colName = col.getColumnName();
+        if (parentOp.getColumnExprMap().containsKey(colName) && candidates.contains(colName)) {
+          partitionCol = colName;
+          break;
+        }
+      }
+
+      if (partitionCol == null) {
+        LOG.debug("Skip grouping-set optimization as there is no feasible column in parent operator {}.",
+            parentOp);
+      }
+
+      return partitionCol;
+    }
+
+    private ReduceSinkOperator createReduceSink(Operator<?> parentOp, String partitionColName,
+        GroupingSetProcessorContext context) {
+      Map<String, ExprNodeDesc> colExprMap = new HashMap<>();
+      List<ExprNodeDesc> valueColumns = new ArrayList<>();
+      List<String> valueColumnNames = new ArrayList<>();
+      List<ColumnInfo> signature = new ArrayList<>();
+      List<ExprNodeDesc> partCols = new ArrayList<>();
+
+      for (ColumnInfo pColInfo: parentOp.getSchema().getSignature()) {
+        ColumnInfo cColInfo = new ColumnInfo(pColInfo);
+        String pColName = pColInfo.getInternalName();
+        valueColumnNames.add(pColName);
+
+        String cColName = Utilities.ReduceField.VALUE + "." + pColName;
+        cColInfo.setInternalName(cColName);
+        signature.add(cColInfo);
+
+        ExprNodeDesc valueExpr = new ExprNodeColumnDesc(pColInfo);
+        valueColumns.add(valueExpr);
+        colExprMap.put(cColName, valueExpr);
+
+        if (pColName.equals(partitionColName)) {
+          partCols.add(valueExpr);
+        }
+      }
+
+      List<FieldSchema> valueFields =
+          PlanUtils.getFieldSchemasFromColumnList(valueColumns, valueColumnNames, 0, "");
+      TableDesc valueTable = PlanUtils.getReduceValueTableDesc(valueFields);
+
+      TableDesc keyTable = PlanUtils.getReduceKeyTableDesc(new ArrayList<>(), "", "");
+
+      List<ExprNodeDesc> keyColumns = new ArrayList<>();
```

Comment:
I could be wrong. Should we also use the partition key as a sort key to optimize the hash table utilization? I see we already achieved a good result with this implementation, so I might be wrong.
If we do that, I wonder if we can generalize the utility to insert REDUCE_SINK + SEL somewhere because it is a useful conversion...

Reply:
I agree that inserting RS + SEL might be useful optimization if we can find the place where heavy sort happens. However, I'm not sure whether HashTable can fully take advantage of pre-sorting.
I'll test your idea in our cluster and share the result later on.

Reply:
@okumin

I tested your idea using 10TB TPC-DS dataset. I used query18, query22, query67 and set `hive.optimize.grouping.set.threshold` to 1,000,000,000. I ran each query 3 times.

||query18|query22|query67|
|---|---|---|---|
|# GBY input rows estimation|382,977,011|258,698,384|5,279,977,323|
|# GBY output rows estimation|1,914,885,055|1,293,491,920|47,519,795,907|
|Query execution time(current impl.) |27.575, 24.089, 23.265|14.827, 13.925, 12.959|426.62, 419.793, 414.403|
|Query execution time(with additional sort)|27.636, 24.836, 24.054|16.531, 15.729, 15.624|318.425, 323.733, 336.545|

There is almost no change in the query execution time of relatively small GroupBys (query18 and query22). However, it appears that a large GroupBy benefits from the additional sort. Therefore, I conclude that adding a sort before large GroupBys is worthwhile. The latest commit(efe982c9) includes an implementation of additional sort that I used for the test.